doc/arch/adr-002-geoprocessing-caching.md

```diff
@@ -0,0 +1,103 @@
+
+### How will the items be identified, cached, and retrieved?
+
+Each geoprocessing run can be identified by two things: 1) input shape, and 2) JSON package sent to geoprocessing service. Because the input shape is a WKAoI, we can use it's table name and id as the identifier, for example `boundary_huc08__1`. The JSON package can be identified by the key in [`settings.GEOP['json']`](https://github.com/WikiWatershed/model-my-watershed/blob/a49cc115d18d2c67cbd0721cf62faa08e98d13fc/src/mmw/mmw/settings/base.py#L387), such as `nlcd_streams`, `nlcd_soils`, `slope`, etc.
+
+We can have a table structure as follows:
+
+| Column     | Type                     | Modifiers |
+|------------|--------------------------|-----------|
+| name       | character varying(255)   | not null  |
+| type       | character varying(255)   | not null  |
+| created_at | timestamp with time zone | not null  |
+
+Where `name` is the name of the shape, for example `boundary_huc08__1`, `type` is the type of cache, for example `nlcd_streams`, and `created_at` is a timestamp of when the cache was created.
+
+When the user requests a WKAoI's analysis, it's table name and id will be passed in, rather than a GeoJSON of the entire shape. The geoprocessing submodule will take the shape name and processing type, and check if a corresponding cache exists. If so, it will immediately return those results. Else, it will run the full calculations and save the results before returning them. The geoprocessing submodule will have the ability to take GeoJSONs in lieu of shape name, in which case the results will not be cached. It will also take an optional parameter which will force recalculation even if nothing has changed.
+
+This caching should be abstracted from the Django Views and Celery Tasks that call the geoprocessing submodule.
+
+### How will the cache be invalidated?
+
+Since every cache entry is tagged with its type, if a certain raster is updated, we can remove all related cache entries. For example, if there are updates to `us-percent-slope-30m-epsg5070` raster, which is used in the [`nlcd_slope`](https://github.com/WikiWatershed/model-my-watershed/blob/a49cc115d18d2c67cbd0721cf62faa08e98d13fc/src/mmw/mmw/settings/base.py#L449) and [`slope`](https://github.com/WikiWatershed/model-my-watershed/blob/a49cc115d18d2c67cbd0721cf62faa08e98d13fc/src/mmw/mmw/settings/base.py#L462) requests, we could simply run
+
+```sql
+DELETE FROM geop_cache
+WHERE type IN ('nlcd_slope', 'slope');
+```
+
+If we'd like to delete all caches before a certain date, for example a deployment, we could do something like
+
+```sql
+DELETE FROM geop_cache
+WHERE created_at < '2017-04-25';
+```
+
+To refresh a specific shape we could do
+
+```sql
+DELETE FROM geop_cache
+WHERE name = 'boundary_huc08__1';
+```
+
+And finally, to delete the entire cache, we can just do
+
+```sql
+DELETE FROM geop_cache;
```

Comment:
Agreed. This was more a demonstration of how the proposed implementation supported various invalidation requirements, but if we go with the Django cache module this may be handled through that and no longer apply. Implementation details shouldn't be in an ADR. Will update and remove.